File: Libro de Flutter/28-proyecto-biblia/02_buscador.dart
// ============================================================
// MÓDULO 28 — Proyecto Biblia (Dart consola)
// Archivo: 02_buscador.dart
//
// DEPENDENCIA: sqlite3: ^2.x.x en pubspec.yaml
//              path: ^1.9.0 en pubspec.yaml
//
// Motor de búsqueda de texto para la Biblia RV60.
// Permite buscar palabras o frases en todo el texto bíblico
// con paginación, estadísticas y filtro por AT/NT.
//
// Ejecutar con:
//   dart run lib/02_buscador.dart
// ============================================================

import 'dart:io';
import 'package:path/path.dart' as path;
import 'package:sqlite3/sqlite3.dart';

/// Expresión regular para limpiar marcas Strong del texto.
final RegExp _strongRegex = RegExp(r'<[^>]+>');

/// Número de resultados por página en la paginación.
const int _resultadosPorPagina = 10;

/// Límite máximo del Antiguo Testamento (libros 1–39).
const int _limiteAT = 39;

void main() async {
  // Calcular ruta absoluta a la BD
  final String scriptDir = path.dirname(Platform.script.toFilePath());
  final String dbPath =
      path.normalize(path.join(scriptDir, '..', 'datos', 'biblia_rv60.sqlite3'));

  if (!File(dbPath).existsSync()) {
    print('ERROR: Base de datos no encontrada en:\n  $dbPath');
    exit(1);
  }

  late final Database db;
  try {
    db = sqlite3.open(dbPath, mode: OpenMode.readOnly);
  } on SqliteException catch (e) {
    print('ERROR al abrir la BD: ${e.message}');
    exit(1);
  }

  try {
    await _menuBuscador(db);
  } finally {
    db.dispose();
  }
}

// ============================================================
// MENÚ DEL BUSCADOR
// ============================================================

/// Menú principal del buscador con cuatro modos de búsqueda.
Future<void> _menuBuscador(Database db) async {
  while (true) {
    print('\n╔═══════════════════════════════════╗');
    print('║  BIBLIA RV60 — Buscador de texto  ║');
    print('╠═══════════════════════════════════╣');
    print('║  1. Buscar en toda la Biblia       ║');
    print('║  2. Buscar solo en el AT           ║');
    print('║  3. Buscar solo en el NT           ║');
    print('║  4. Salir                          ║');
    print('╚═══════════════════════════════════╝');
    stdout.write('Opción: ');

    final opcion = stdin.readLineSync()?.trim() ?? '';
    if (opcion == '4') {
      print('\n¡Hasta luego!\n');
      return;
    }

    if (!['1', '2', '3'].contains(opcion)) {
      print('Opción no válida.');
      continue;
    }

    stdout.write('\nTérmino de búsqueda: ');
    final termino = stdin.readLineSync()?.trim() ?? '';
    if (termino.isEmpty) continue;

    // Determinar el filtro de testamento
    int? libroMin;
    int? libroMax;
    String etiquetaTestamento = 'toda la Biblia';
    if (opcion == '2') {
      libroMax = _limiteAT;
      etiquetaTestamento = 'el Antiguo Testamento';
    } else if (opcion == '3') {
      libroMin = _limiteAT + 1;
      etiquetaTestamento = 'el Nuevo Testamento';
    }

    await _ejecutarBusqueda(db, termino, etiquetaTestamento, libroMin, libroMax);
  }
}

// ============================================================
// LÓGICA DE BÚSQUEDA Y PAGINACIÓN
// ============================================================

/// Ejecuta la búsqueda y muestra los resultados con paginación.
///
/// [termino]            → palabra o frase a buscar
/// [etiquetaTestamento] → texto para el encabezado ("AT", "NT", "toda la Biblia")
/// [libroMin]           → número de libro mínimo (null = sin límite inferior)
/// [libroMax]           → número de libro máximo (null = sin límite superior)
Future<void> _ejecutarBusqueda(
  Database db,
  String termino,
  String etiquetaTestamento, [
  int? libroMin,
  int? libroMax,
]) async {
  // Construir la cláusula WHERE dinámicamente
  final condiciones = <String>['v.text LIKE ?'];
  final parametros = <Object?>['%$termino%'];

  if (libroMin != null) {
    condiciones.add('v.book_number >= ?');
    parametros.add(libroMin);
  }
  if (libroMax != null) {
    condiciones.add('v.book_number <= ?');
    parametros.add(libroMax);
  }

  final where = condiciones.join(' AND ');

  // Obtener el total de resultados para estadísticas
  final totalResultado = db.select(
    'SELECT COUNT(*) AS n FROM verses v WHERE $where',
    parametros,
  );
  final totalVersiculos = totalResultado.first['n'] as int;

  if (totalVersiculos == 0) {
    print('\nNo se encontró "$termino" en $etiquetaTestamento.');
    return;
  }

  // Contar en cuántos libros distintos aparece el término
  final librosResultado = db.select(
    'SELECT COUNT(DISTINCT v.book_number) AS n FROM verses v WHERE $where',
    parametros,
  );
  final totalLibros = librosResultado.first['n'] as int;

  // Mostrar encabezado de estadísticas
  print('\n═══════════════════════════════════════════════');
  print('Búsqueda: "$termino" en $etiquetaTestamento');
  print('Resultados: $totalVersiculos versículos en $totalLibros libros');
  print('═══════════════════════════════════════════════');

  // Paginar resultados
  int offset = 0;
  while (offset < totalVersiculos) {
    final pagina = _obtenerPagina(db, where, parametros, offset);
    _mostrarPagina(pagina, offset + 1, termino);

    offset += _resultadosPorPagina;
    if (offset >= totalVersiculos) break;

    final restantes = totalVersiculos - offset;
    print('\n[Mostrando ${min(offset, totalVersiculos)} de $totalVersiculos]');
    stdout.write('¿Ver más resultados? ($restantes restantes) [s/N]: ');
    final respuesta = stdin.readLineSync()?.trim().toLowerCase() ?? '';
    if (respuesta != 's') break;
  }

  // Mostrar resumen por libros
  _mostrarResumenPorLibros(db, where, parametros);
}

/// Obtiene una página de resultados con los datos completos de referencia.
List<Row> _obtenerPagina(
  Database db,
  String where,
  List<Object?> parametros,
  int offset,
) {
  return db.select(
    'SELECT b.long_name, v.chapter, v.verse, v.text '
    'FROM verses v '
    'JOIN books b ON v.book_number = b.book_number '
    'WHERE $where '
    'ORDER BY v.book_number, v.chapter, v.verse '
    'LIMIT ? OFFSET ?',
    [...parametros, _resultadosPorPagina, offset],
  );
}

/// Muestra una página de resultados con el término resaltado por contexto.
/// [inicio] → número del primer versículo en esta página (para numeración).
void _mostrarPagina(List<Row> pagina, int inicio, String termino) {
  for (int i = 0; i < pagina.length; i++) {
    final fila = pagina[i];
    final num = (inicio + i).toString().padLeft(4);
    final referencia =
        '${fila['long_name']} ${fila['chapter']}:${fila['verse']}';
    final texto = _limpiarMarcas(fila['text'].toString());
    print('\n$num. $referencia');
    print('     $texto');
  }
}

/// Muestra los 5 libros con más apariciones del término buscado.
void _mostrarResumenPorLibros(
  Database db,
  String where,
  List<Object?> parametros,
) {
  final resultado = db.select(
    'SELECT b.long_name, COUNT(*) AS ocurrencias '
    'FROM verses v '
    'JOIN books b ON v.book_number = b.book_number '
    'WHERE $where '
    'GROUP BY v.book_number '
    'ORDER BY ocurrencias DESC '
    'LIMIT 5',
    parametros,
  );

  print('\n─── Top 5 libros con más ocurrencias ───');
  for (int i = 0; i < resultado.length; i++) {
    final fila = resultado[i];
    final pos = (i + 1).toString().padLeft(2);
    final nombre = fila['long_name'].toString().padRight(20);
    print('  $pos. $nombre ${fila['ocurrencias']} veces');
  }
}

// ============================================================
// UTILIDADES
// ============================================================

/// Elimina marcas Strong y etiquetas del texto del versículo.
String _limpiarMarcas(String texto) =>
    texto.replaceAll(_strongRegex, '').trim();

/// Devuelve el mínimo entre dos enteros.
int min(int a, int b) => a < b ? a : b;

File: Libro de JavaScript/18-proyecto-biblia-consola/02_buscador.js
// =============================================================================
// ARCHIVO: 02_buscador.js
// TEMA: Buscador de versículos con paginación y filtro por testamento
// =============================================================================
// DEPENDENCIA: npm install better-sqlite3

const Database = require('better-sqlite3');
const readline = require('readline');
const path = require('path');

const DB_PATH = path.join(__dirname, '..', '..', 'datos', 'biblia_rv60.sqlite3');
const db = new Database(DB_PATH, { readonly: true });

// Constantes para los rangos de testamento
const AT_MAX = 39; // Libros 1-39: Antiguo Testamento
const NT_MIN = 40; // Libros 40-66: Nuevo Testamento

/** Tamaño de página para la paginación de resultados */
const TAMANIO_PAGINA = 10;

/**
 * Elimina marcas Strong del texto bíblico.
 * @param {string} texto - Texto con posibles marcas <S>NNNN</S>
 * @returns {string} Texto limpio
 */
const limpiarMarcas = texto =>
  texto.replace(/<S>\d+<\/S>/g, '').replace(/\s{2,}/g, ' ').trim();

/**
 * Busca versículos que contengan la palabra clave dada.
 * Opcionalmente filtra por testamento (AT, NT, o ambos).
 * @param {string} palabra - Término de búsqueda (SQL LIKE)
 * @param {'AT'|'NT'|'TODO'} testamento - Filtro de testamento
 * @returns {Array} Lista de resultados con referencia y texto limpio
 */
function buscar(palabra, testamento = 'TODO') {
  let condicionTestamento = '';
  if (testamento === 'AT') condicionTestamento = `AND v.book_number <= ${AT_MAX}`;
  if (testamento === 'NT') condicionTestamento = `AND v.book_number >= ${NT_MIN}`;

  const filas = db
    .prepare(
      `SELECT v.book_number, v.chapter, v.verse, v.text, b.long_name, b.short_name
       FROM verses v
       JOIN books b ON v.book_number = b.book_number
       WHERE v.text LIKE ? ${condicionTestamento}
       ORDER BY v.book_number, v.chapter, v.verse`
    )
    .all(`%${palabra}%`);

  return filas.map(f => ({
    referencia: `${f.long_name} ${f.chapter}:${f.verse}`,
    texto: limpiarMarcas(f.text),
    libro: f.long_name,
    bookNum: f.book_number,
  }));
}

/**
 * Calcula estadísticas sobre un conjunto de resultados de búsqueda.
 * Devuelve el total de versículos y el número de libros distintos involucrados.
 * @param {Array} resultados - Lista de resultados de buscar()
 * @returns {object} { total, librosUnicos }
 */
function calcularEstadisticas(resultados) {
  const libros = new Set(resultados.map(r => r.libro));
  return {
    total: resultados.length,
    librosUnicos: libros.size,
  };
}

/**
 * Imprime una página de resultados en consola, con número de página.
 * @param {Array} resultados - Todos los resultados de la búsqueda
 * @param {number} pagina - Número de página actual (empezando en 0)
 */
function imprimirPagina(resultados, pagina) {
  const inicio = pagina * TAMANIO_PAGINA;
  const fin = Math.min(inicio + TAMANIO_PAGINA, resultados.length);
  const pagina_actual = resultados.slice(inicio, fin);

  console.log(`\n  — Resultados ${inicio + 1}-${fin} de ${resultados.length} —\n`);
  pagina_actual.forEach((r, i) => {
    console.log(`  ${String(inicio + i + 1).padStart(4)}. ${r.referencia}`);
    console.log(`        ${r.texto.substring(0, 100)}${r.texto.length > 100 ? '...' : ''}`);
  });
}

// ─── Interfaz de usuario ─────────────────────────────────────────────────────

const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
const preguntar = msg => new Promise(res => rl.question(msg, res));

/**
 * Flujo principal del buscador: solicita término, testamento,
 * muestra estadísticas y permite paginar resultados.
 */
async function iniciarBuscador() {
  console.log('\n╔══════════════════════════════════╗');
  console.log('║   BIBLIA RV60 — Buscador         ║');
  console.log('╚══════════════════════════════════╝\n');

  let continuar = true;
  while (continuar) {
    const palabra = await preguntar('Buscar palabra (o "salir"): ');
    if (palabra.toLowerCase() === 'salir') break;
    if (!palabra.trim()) continue;

    console.log('  Testamento: 1=Todo  2=Solo AT  3=Solo NT');
    const opT = await preguntar('  Opción (1-3): ');
    const mapa = { '1': 'TODO', '2': 'AT', '3': 'NT' };
    const testamento = mapa[opT.trim()] || 'TODO';

    console.log('\n  Buscando...');
    const resultados = buscar(palabra.trim(), testamento);

    if (resultados.length === 0) {
      console.log(`\n  ✗ No se encontraron resultados para "${palabra}".`);
      continue;
    }

    const stats = calcularEstadisticas(resultados);
    console.log(`\n  ✓ ${stats.total} versículos en ${stats.librosUnicos} libro(s)`);

    let pagina = 0;
    const totalPaginas = Math.ceil(resultados.length / TAMANIO_PAGINA);

    imprimirPagina(resultados, pagina);

    while (pagina < totalPaginas - 1) {
      const mas = await preguntar('\n  [ENTER] ver más  |  [s] nueva búsqueda: ');
      if (mas.toLowerCase() === 's') break;
      pagina++;
      imprimirPagina(resultados, pagina);
    }

    if (pagina >= totalPaginas - 1) {
      console.log('\n  — Fin de los resultados —');
    }
  }

  rl.close();
  db.close();
  console.log('\nHasta luego.\n');
}

iniciarBuscador().catch(err => {
  console.error('Error:', err.message);
  db.close();
  process.exit(1);
});
